### ISSUE TITLE
Operator: Repackage Installation to allow choice of technical implementations

### ISSUE THREAD

**Issue by LesterThomas** [tmforum-oda/oda-canvas#364]
## Description

The current Canvas installation is a helm-of-helm charts that installs a default set of operators. The [Canvas Helm chart](https://github.com/tmforum-oda/oda-canvas/tree/master/charts/canvas-oda) references a dependency of separate helm charts for each operator.

The Kong and APISIX operators can be installed over-the-top to add an API Gateway (and reconfigure the Istio service Mesh). 

This issue is to make a single Canvas install a declarative process with choice of technology implementation. The canvas helm-of-helm values file should contain a list of possible operator for each class of operator. Only one choice for each operator should be enabled. A helm pre-install hook could check that a values file is valid before attempting an install.

The values file would look something like:

```helm
API-Management:
  apiOperatorIstio:
    enabled: true
  apiOperatorKong:
    enabled: false
  apiOperatorApisix:
    enabled: false
  apiOperatorApig:
    enabled: false
Identity-Config:
  identityConfigKeycloak:
    enabled: true
  identityConfigEntraID:
    enabled: false
```

**Comment by LesterThomas** [comment]
@RJ-acc - At the moment we have an `entrypoint.sh` file for the component operator that runs either the `apiOperatorIstiowithKong.py` or `apiOperatorIstiowithApisix.py` or `apiOperatorIstio.py` file as part of component operator. I propose that we split this functionality from the API Operator and package the `apiOperatorIstiowithKong.py` file with the Kong operator, the `apiOperatorIstiowithApisix.py` with the APISIX operator and create a corresponding Istio operator with the `apiOperatorIstio.py` file.

This would result in the installation requiring the Component operator and then one of the Kong/APISIX/Istio operators (i.e. these become peers of each other).

What do you think?

**Comment by RJ-acc** [comment]
@LesterThomas I completely agree with you for this. But we need to sort Kopf Operator Priority issue if we want to run them in this way. In first version of API gateway operators (Kong and Apisix), they were running in separate PODs, but issues were observed.

**Comment by LesterThomas** [comment]
> @LesterThomas I completely agree with you for this. But we need to sort Kopf Operator Priority issue if we want to run them in this way. In first version of API gateway operators (Kong and Apisix), they were running in separate PODs, but issues were observed.

I have done some experimentation with Kopf and have a workaround for the issue. The issue was that Kopf caused repeated updates to a Kubernetes resource if two different operators in two different pods had handlers for the same resource (it works fine if you package those two operators into the same image).

I moved the ‘updateAPIReady’ and ‘updateAPIStatus’ handlers from the component operator to the api operator. These two handlers were triggered by the status of the ExposedAPI resource (they then patch the parent component status). Kopf has issues with the component operator being triggered by updates to the API resource - it works fine if these updates are packaged with the api operator.

By moving these handlers, I was able to package the component operator and api operator into two separate containers images running in two pods.
